Sources/PenguinStructures/AnyValue.swift

```diff
@@ -0,0 +1,218 @@
+    init(_ address: UnsafeRawPointer) {
+      self.address = .init(mutating: address)
+    }
+    /// The address of the `Any` accessed by `self`
+    private let address: UnsafeMutableRawPointer
+
+    /// A pointer to the stored instance of type `T`.
+    public var pointer: UnsafeMutablePointer<T> {
+      if MemoryLayout<T>.fitsExistentialInlineBuffer {
+        return address.assumingMemoryBound(to: T.self)
+      }
+      else {
+        let boxBase = address.assumingMemoryBound(to: BoxBase.self).pointee
+        let box = unsafeDowncast(boxBase, to: Box<T>.self)
+        return withUnsafeMutablePointer(to: &box.value) { $0 }
+      }
+    }
+  }
+
+  /// Iff `storedType != T.self`, traps with an appropriate error message.
+  private func typeCheck<T>(_: Type<T>) {
+    if storedType != T.self { typeCheckFailure(T.self) }
+  }
+
+  /// Traps with an appropriate error message assuming that `storedType != T.self`.
+  @inline(never)
+  private func typeCheckFailure(_ expectedType: Any.Type) {
+    fatalError("stored type \(storedType) != \(expectedType)")
+  }
+
+  /// Accesses the `T` stored in `self`.
+  ///
+  /// - Requires: `storedType == T.self`.
+  @inline(__always) // Compiler likes to skip inlining this otherwise.
+  public subscript<T>(_: Type<T>) -> T {
+    get {
+      defer { _fixLifetime(self) }
+      typeCheck(Type<T>())
+      return pointer(toStored: Type<T>()).pointee
+    }
+    _modify {
+      typeCheck(Type<T>())
+      yield &mutablePointer(toStored: Type<T>()).pointee
+      _fixLifetime(self)
+    }
+  }
+
+  /// Unsafely accesses the `T` stored in `self`.
+  ///
+  /// - Requires: `storedType == T.self`.
+  @inline(__always) // Compiler likes to skip inlining this otherwise.
+  public subscript<T>(unsafelyAssuming _: Type<T>) -> T {
+    get {
+      defer { _fixLifetime(self) }
+      return pointer(toStored: Type<T>()).pointee
+    }
+    _modify {
+      yield &mutablePointer(toStored: Type<T>()).pointee
+      _fixLifetime(self)
+    }
+  }
+
+  /// Stores `x` in `self`.
+  ///
+  /// This may be more efficient than `self = AnyValue(x)` because it uses the same allocated buffer
+  /// when possible for large types.
+  public mutating func store<T>(_ x: T) {
+    defer { _fixLifetime(self) }
+    if storedType == T.self { mutablePointer(toStored: Type<T>()).pointee = x }
+    else { self = .init(x) }
+  }
+
+  /// The stored value.
+  ///
+  /// This property can be useful for interoperability with the rest of Swift, especially when you
+  /// don't know the full dynamic type of the stored value.
+  public var asAny: Any {
+    if directlyStoredType != BoxBase.self { return storage }
+    return self[unsafelyAssuming: Type<BoxBase>()].asAny
+  }
+
+  /// If the stored value is boxed or is an object, returns the ID of the box or object; returns
+  /// `nil` otherwise.
+  ///
+  /// Used by tests.
+  internal var boxOrObjectID_: ObjectIdentifier? {
```

Comment:
Why is this underscore-suffixed? (e.g. vs underscore prefixed or any other typical convention.)

> Note: I do like the pattern of performing `@testable import`s only to expose test utilities and using the public API by default in test code. (We might want to document that as the recommended pattern somewhere.)

Reply:
underscore-prefixed means something specific that doesn't apply here.  Pretty sure we've been over that.  

I don't care what convention we use for such things; I just picked something.  Maybe `_testable`-suffixed would be better?